src/Servers/Kestrel/Transport.NamedPipes/test/NamedPipeConnectionListenerTests.cs

```diff
@@ -163,4 +163,56 @@ public async Task BindAsync_ListenersSharePort_DisposeFirstListener_Success()
         await using var connectionListener2 = await NamedPipeTestHelpers.CreateConnectionListenerFactory(LoggerFactory, pipeName: pipeName);
         Assert.Equal(connectionListener1.EndPoint, connectionListener2.EndPoint);
     }
+
+    [Fact]
+    public async Task CompleteTransport_BeforeClientHasReadLastBytes_DontLooseData()
+    {
+        var options = new NamedPipeTransportOptions();
+        await using var connectionListener = await NamedPipeTestHelpers.CreateConnectionListenerFactory(LoggerFactory, options: options);
+
+        var clientStream = NamedPipeTestHelpers.CreateClientStream(connectionListener.EndPoint);
+        var connectTask = clientStream.ConnectAsync();
+
+        var connection = await connectionListener.AcceptAsync();
+        await connection.Transport.Output.WriteAsync(new byte[] { 41 });
+        byte[] readBuffer = new byte[1];
+        var readAmount = await clientStream.ReadAsync(readBuffer);
+        Assert.Equal(1, readAmount);
+        Assert.Equal(41, readBuffer[0]);
+        await connection.Transport.Output.WriteAsync(new byte[] { 42 });
+        // complete transport and wait for a reasonable amount of time
+        // actual stream disconnection should happen after the last client read happens
+        connection.Transport.Output.Complete();
+        await Task.Delay(500);
+        readAmount = await clientStream.ReadAsync(readBuffer);
+        Assert.Equal(1, readAmount);
+        Assert.Equal(42, readBuffer[0]);
+    }
+
+    [Fact]
+    public async Task DisposeAsync_BeforeClientHasReadLastBytes_DontLooseData()
+    {
+        var options = new NamedPipeTransportOptions();
+        await using var connectionListener = await NamedPipeTestHelpers.CreateConnectionListenerFactory(LoggerFactory, options: options);
+
+        var clientStream = NamedPipeTestHelpers.CreateClientStream(connectionListener.EndPoint);
+        var connectTask = clientStream.ConnectAsync();
+
+        var connection = await connectionListener.AcceptAsync();
+        await connection.Transport.Output.WriteAsync(new byte[] { 41 });
+        byte[] readBuffer = new byte[1];
+        var readAmount = await clientStream.ReadAsync(readBuffer);
+        Assert.Equal(1, readAmount);
+        Assert.Equal(41, readBuffer[0]);
+        await connection.Transport.Output.WriteAsync(new byte[] { 42 });
+        // trigger dispose and wait for a reasonable amount of time
+        // actual stream disconnection should happen after the last client read happens
+        var disposeTask = connection.DisposeAsync();
+        await Task.Delay(500);
```

Comment:
What is the delay here waiting for?

Reply:
That is actually the same as above: we want the sending task not to complete until the client has read the stream to end to avoid disposing the stream too early.